### PR TITLE
Review images documentation.

### PR DESCRIPTION
**Pull request by seonmej**

### CODE REVIEW

File: VEEPortingGuide/uiImageRaw.rst [hunk 1]
```diff
@@ -10,35 +10,44 @@ The Image Engine makes the distinction between:
...
+MicroEJ Format: Display
+=======================
 
-Several MicroEJ format encodings are available. Some encodings may be directly managed by the display driver. Refers to the platform specification to retrieve the list of better formats.
+See :ref:`section_image_display_output`.
 
-Advantages:
+The display can hold a pixel encoding which is not standard (see :ref:`display_pixel_structure`).
+The MicroEJ format can be customized to encode the pixel in same encoding than display.
+The number of bits-per-pixels and the pixel bits organisation is asked during the MicroEJ format generation and when the ``drawImage`` algorithms are running.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:34]:
```suggestion
The number of bits per pixel and the pixel bit organization is asked during the MicroEJ format generation and when the ``drawImage`` algorithms are running.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 2]
```diff
@@ -10,35 +10,44 @@ The Image Engine makes the distinction between:
...
+If the image to encode contains some transparent pixels, the output file will embed the transparency according to the display’s implementation capacity.
+When all pixels are fully opaque, no extra information will be stored in the output file in order to free up some memory space.
 
-* The pixels layout and bits format are standard, so it is easy to manipulate these images on the C-side.
-* Drawing an image is very fast when the display driver recognizes the format (with or without transparency).
-* Supports or not the alpha encoding: select the better format according to the image to encode.
+.. note:: From Image Engine point of view, the format stays a MicroEJ format, readable by the Image Renderer.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:38]:
```suggestion
.. note:: From the Image Engine point of view, the format stays a MicroEJ format, readable by the Image Renderer.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 3]
```diff
@@ -10,35 +10,44 @@ The Image Engine makes the distinction between:
...
 
-* No compression: the image size in bytes is proportional to the number of pixels, the transparency, and the number of bits-per-pixel. 
-* Slower than ``display`` format when the display driver does not recognize the  format: a pixel conversion is required at runtime.
+MicroEJ Format: Standard
+========================
 
-This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride etc. The required memory also depends on number of bits-per-pixels of MicroEJ format:
+See :ref:`section_image_standard_output`.
 
+This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride etc.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:49]:
```suggestion
This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride, etc.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 4]
```diff
@@ -10,35 +10,44 @@ The Image Engine makes the distinction between:
...
+The required memory also depends on the number of bits-per-pixels of MicroEJ format:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:50]:
```suggestion
The required memory also depends on the number of bits per pixels of the MicroEJ format:
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 5]
```diff
@@ -92,9 +90,14 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
--  RGB565: 16 bits format, 5 or 6 per color. Image is always fully
-   opaque.
+-  RGB888: 24-bit format, 8 per color.
+   ::
 
+      u32 convertARGB8888toRAWFormat(u32 c){
+          return c & 0xffffff;
+      }
+
+-  RGB565: 16-bit format, 5 or 6 per color.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:100]:
```suggestion
-  RGB565: 16-bit format, 5 for red, 6 for green, 5 for blue.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 6]
```diff
@@ -105,71 +108,51 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
    ::
 
       u32 convertARGB8888toRAWFormat(u32 c){
           return (0xff - (toGrayscale(c) & 0xff)) / 0xff;
       }
 
--  C4: 4 bits format with grayscale conversion. Image is always fully
-   opaque.
-
+The pixels order follows this rule:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:139]:
```suggestion
The pixel order follows this rule:
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 7]
```diff
@@ -105,71 +108,51 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
-      }
+MicroEJ Format: Grayscale
+=========================
 
--  C1: 1 bit format with grayscale conversion. Image is always fully
-   opaque.
+See :ref:`section_image_grayscale_output`.
 
-   ::
+This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride etc.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:149]:
```suggestion
This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride, etc.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 8]
```diff
@@ -105,71 +108,51 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
+The required memory also depends on the number of bits-per-pixels of MicroEJ format:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:150]:
```suggestion
The required memory also depends on the number of bits per pixels of the MicroEJ format:
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 9]
```diff
@@ -201,40 +182,49 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
+-  C1: 1 bit with grayscale conversion.
+   ::
 
-.. note:: From Image Engine point of view, the format stays a MicroEJ format, readable by the Image Renderer.
+      u32 convertARGB8888toRAWFormat(u32 c){
+          return (toGrayscale(c) & 0xff) / 0xff;
+      }
 
-Advantages:
+The pixels order follows this rule:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:206]:
```suggestion
The pixel order follows this rule:
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 10]
```diff
@@ -201,40 +182,49 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
-* No compression: the image size in bytes is proportional to the number of pixels. The required memory is similar to :ref:`section_image_standard_raw`.
+See :ref:`section_image_rle_output`.
 
 .. _section_image_gpu_raw:
 
 MicroEJ Format: GPU
 ===================
 
-The MicroEJ format may be customized to be platform's GPU compatible. It can be extanded by one or several restrictions on the pixels array: 
+The MicroEJ format may be customized to be platform's GPU compatible. It can be extended by one or several restrictions on the pixels array:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:221]:
```suggestion
The MicroEJ format may be customized to be compatible with the platform's GPU. It can be extended by one or several restrictions on the pixels array: 
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 11]
```diff
@@ -201,40 +182,49 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
 MicroEJ Format: GPU
 ===================
 
-The MicroEJ format may be customized to be platform's GPU compatible. It can be extanded by one or several restrictions on the pixels array: 
+The MicroEJ format may be customized to be platform's GPU compatible. It can be extended by one or several restrictions on the pixels array: 
 
 * Its start address has to be aligned on a higher value than the number of bits-per-pixels. 
 * A padding has to be added after each line (row stride).
-* The MicroEJ format can hold a platform dependant header, located between MicroEJ format header (start of file) and pixels array. The MicroEJ format is designed to let the platform encodes and decodes this additional header. For Image Engine software algorithms, this header is useless and never used. 
+* The MicroEJ format can hold a platform dependant header, located between MicroEJ format header (start of file) and pixels array.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:225]:
```suggestion
* The MicroEJ format can hold a platform-dependent header, located between the MicroEJ format header (start of file) and the pixel array.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 12]
```diff
@@ -201,40 +182,49 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
+  The MicroEJ format is designed to let the platform encodes and decodes this additional header.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:226]:
```suggestion
  The MicroEJ format is designed to let the platform encode and decode this additional header.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 13]
```diff
@@ -201,40 +182,49 @@ Select one the following format to use a generic format among this list: ``ARGB8
...
+  For Image Engine software algorithms, this header is useless and never used.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:227]:
```suggestion
  For Image Engine software algorithms, this header is not needed and never used.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 14]
```diff
@@ -246,64 +236,32 @@ Advantages:
 
 Disadvantages:
 
-* No compression: the image size in bytes is proportional to the number of pixels. The required memory is similar to :ref:`section_image_standard_raw` when there is no custom header. 
-
-When MicroEJ format holds another header (called ``custom_header``), the required memory depends is:
+* No compression: the image size in bytes is proportional to the number of pixels.
+  The required memory is similar to :ref:`section_image_standard_raw` when there is no custom header. 
 
+When MicroEJ format holds another header (called ``custom_header``), the required memory is:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:242]:
```suggestion
When the MicroEJ format holds another header (called ``custom_header``), the required memory is:
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 15]
```diff
@@ -246,64 +236,32 @@ Advantages:
...
 ::
 
       required_memory = header + custom_header + (image_width * image_height) * bpp / 8;
 
-The row stride allows to add some padding at the end of each line in order to start next line at an address with a specific memory alignment; it is often required by hardware accelerators (GPU). The row stride is by default a value in relation with the image width: ``row_stride_in_bytes = image_width * bpp / 8``. It can be customized at image buffer creation thanks to the Low Level API ``LLUI_DISPLAY_IMPL_getNewImageStrideInBytes``.  The required RAM memory becomes:
-
+The row stride allows to add some padding at the end of each line in order to start next line at an address with a specific memory alignment; it is often required by hardware accelerators (GPU).
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:247]:
```suggestion
The row stride allows to add some padding at the end of each line in order to start the next line at an address with a specific memory alignment; it is often required by hardware accelerators (GPU).
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 16]
```diff
@@ -246,64 +236,32 @@ Advantages:
...
+The row stride is by default a value in relation with the image width: ``row_stride_in_bytes = image_width * bpp / 8``.
+It can be customized at image buffer creation thanks to the Low Level API ``LLUI_DISPLAY_IMPL_getNewImageStrideInBytes``.
+The required RAM memory becomes:
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:250]:
```suggestion
The required memory becomes:
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 17]
```diff
@@ -246,64 +236,32 @@ Advantages:
...
-
-The file format is quite similar to :ref:`section_image_standard_raw`.
-
 .. _section_image_binary_raw:
 
 Binary Format
 =============
 
-This format is not compatible with the Image Renderer and by MicroUI. It is can be used by MicroUI addon libraries which provide their own images managements. 
+This format is not compatible with the Image Renderer and by MicroUI.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:263]:
```suggestion
This format is not compatible with the Image Renderer and MicroUI.
```

File: VEEPortingGuide/uiImageRaw.rst [hunk 18]
```diff
@@ -246,64 +236,32 @@ Advantages:
...
+It is can be used by MicroUI addon libraries which provide their own images managements.
```
Comment (microej-sh) [inline VEEPortingGuide/uiImageRaw.rst:264]:
```suggestion
It is can be used by MicroUI addon libraries which provide their own image management procedures.
```

File: ApplicationDeveloperGuide/UI/MicroUI/images.rst [hunk 19]
```diff
@@ -464,10 +426,12 @@ Images Heap
...
-- immutable images which are not byte-addressable, such as images opened with an input stream (i.e. `ResourceImage`_  instances)
-- immutable images which are byte-addressable but converted to a different output format (i.e. `ResourceImage`_  instances)
+- Mutable images (i.e. `BufferedImage`_  instances).
+- Immutable images decoded at runtime, typically a PNG: the heap is used to store the decoded image **and** the runtime decoder's temporary buffers, required during the decoding step.
+  After the decoding step, all the temporary buffers are freed.
+  Note that the size of the temporary buffers depend on the decoder **and** on the original image itself (compression level, pixel encoding, etc.).
+- Immutable images which are not byte-addressable, such as images opened with an input stream (i.e. `ResourceImage`_  instances).
+- Immutable images which are byte-addressable but converted to a different output format (i.e. `ResourceImage`_  instances).
 
 In other words, every image which can not be retrieved using `ej.microui.display.Image.getImage()`_  is saved on the images heap.
```
Comment (microej-sh) [inline ApplicationDeveloperGuide/UI/MicroUI/images.rst:436]:
```suggestion
In other words, every image which cannot be retrieved using `ej.microui.display.Image.getImage()`_  is saved on the image heap.
```

File: VEEPortingGuide/uiImageRaw.rst (hunk 18 above)
Comment (seonmej) [reply VEEPortingGuide/uiImageRaw.rst:264]: `It is can be` also ;)

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 20]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
+
+* The rendering time of a simple shape with the GPU (time in the VG-Lite library + GPU setup time + rendering time) is more important than the direct rendering in software. To enable the hardware rendering for the simple shapes, set the define ``VGLITE_USE_GPU_FOR_SIMPLE_DRAWINGS``  in ``display_configuration.h``.
+* The rendering time of a RGB565 image on a RGB565 buffer without applying an opacity (alpha == 0xff) is more important than than the direct rendering in software (this kind of drawing consists to perform a simple memory copy). To enable the hardware rendering for the RGB565 images, set the define ``VGLITE_USE_GPU_FOR_RGB565_IMAGES``  in ``display_configuration.h``.
+* According to the VG-Lite GPU revision, the transparent images ARGB8888, ARGB1555 and ARGB4444 are not compatible. The old GPU revisions don't render correctly the transparent images because the pre-multiplication of the pixel opacity is not propagated to the pixel color components. To enable the hardware rendering for the transparent images, set the define ``VGLITE_USE_GPU_FOR_TRANSPARENT_IMAGES``  in ``display_configuration.h``. Note this limitation does not concern the A8 and A4 formats.
```
Comment (seonmej) [inline VEEPortingGuide/uiReleaseNotes.rst:287]: Depending on the VG-Lite GPU revision, the transparent images…

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 21]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
+The following table describes the accelerated features:
+
++----------------+------------------------------------------------------+
+| Feature        | Comment                                              |
++================+======================================================+
+| fill rectangle |                                                      |
++----------------+------------------------------------------------------+
+| draw image     | ARGB8888, RGB888, RGB565, ARGB1555, ARGB4444, A8, A4 |
++----------------+------------------------------------------------------+
+| flush (copy)   | copy of data from back buffer to frame buffer        |
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst:262]:
```suggestion
| Fill rectangle |                                                      |
+----------------+------------------------------------------------------+
| Draw image     | ARGB8888, RGB888, RGB565, ARGB1555, ARGB4444, A8, A4 |
+----------------+------------------------------------------------------+
| Flush (copy)   | Copy data from back buffer to frame buffer           |
```

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 22]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
++----------------+------------------------------------------------------+
+
+The following table describes the compatibility versions between the C module and the MicroEJ UI Packs:
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst:265]:
```suggestion
The following table describes the version compatibility between the C module and the MicroEJ UI Packs:
```

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 23]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
+
+.. _com.microej.clibrary.llimpl(display-dma2d): https://repository.microej.com/modules/com/microej/clibrary/llimpl/display-dma2d/
+
+**Vivante VG-Lite**
+
+The VG-Lite C module `com.microej.clibrary.llimpl(microui-vglite)`_ targets the NXP CPU that provide the Vivante VG-Lite accelerator. 
+
+This C module provides some drawing algorithms that are disabled by default. 
+
+* The rendering time of a simple shape with the GPU (time in the VG-Lite library + GPU setup time + rendering time) is more important than the direct rendering in software. To enable the hardware rendering for the simple shapes, set the define ``VGLITE_USE_GPU_FOR_SIMPLE_DRAWINGS``  in ``display_configuration.h``.
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst:285]:
```suggestion
* The rendering time of a simple shape with the GPU (time in the VG-Lite library + GPU setup time + rendering time) is longer than with software rendering. To enable the hardware rendering for simple shapes, uncomment the definition of ``VGLITE_USE_GPU_FOR_SIMPLE_DRAWINGS``  in ``display_configuration.h``.
```

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 24]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
+* The rendering time of a RGB565 image on a RGB565 buffer without applying an opacity (alpha == 0xff) is more important than than the direct rendering in software (this kind of drawing consists to perform a simple memory copy). To enable the hardware rendering for the RGB565 images, set the define ``VGLITE_USE_GPU_FOR_RGB565_IMAGES``  in ``display_configuration.h``.
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst]:
```suggestion
* The rendering time of an RGB565 image into an RGB565 buffer without applying an opacity (alpha == 0xff) is longer than than with software rendering (as this kind of drawing consists in performing a mere memory copy). To enable the hardware rendering for RGB565 images, uncomment the definition of ``VGLITE_USE_GPU_FOR_RGB565_IMAGES``  in ``display_configuration.h``.
```

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 25]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
+| draw thick circle arc       |                                                           |
++-----------------------------+-----------------------------------------------------------+
+| draw thick ellipse          |                                                           |
++-----------------------------+-----------------------------------------------------------+
+| draw flipped image          | see draw image                                            |
++-----------------------------+-----------------------------------------------------------+
+| draw rotated image          | see draw image                                            |
++-----------------------------+-----------------------------------------------------------+
+| draw scaled image           | see draw image                                            |
++-----------------------------+-----------------------------------------------------------+
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst:343]:
```suggestion
+-----------------------------+-----------------------------------------------------------+
| Feature                     | Comment                                                   |
+=============================+===========================================================+
| Draw line                   | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Fill rectangle              | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Draw rounded rectangle      | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Fill rounded rectangle      |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw circle arc             | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Fill circle arc             |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw ellipse arc            | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Fill ellipse arc            |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw ellipse arc            | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Fill ellipse arc            |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw circle                 | Disabled by default (see above)                           |
+-----------------------------+-----------------------------------------------------------+
| Fill circle                 |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw image                  | RGB565, A8, A4 + ARGB8888, ARGB1555, ARGB4444 (see above) |
+-----------------------------+-----------------------------------------------------------+
| Draw thick faded point      | Only with fade <= 1                                       |
+-----------------------------+-----------------------------------------------------------+
| Draw thick faded line       | Only with fade <= 1                                       |
+-----------------------------+-----------------------------------------------------------+
| Draw thick faded circle     | Only with fade <= 1                                       |
+-----------------------------+-----------------------------------------------------------+
| Draw thick faded circle arc | Only with fade <= 1                                       |
+-----------------------------+-----------------------------------------------------------+
| Draw thick faded ellipse    | Only with fade <= 1                                       |
+-----------------------------+-----------------------------------------------------------+
| Draw thick line             |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw thick circle           |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw thick circle arc       |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw thick ellipse          |                                                           |
+-----------------------------+-----------------------------------------------------------+
| Draw flipped image          | See draw image                                            |
+-----------------------------+-----------------------------------------------------------+
| Draw rotated image          | See draw image                                            |
+-----------------------------+-----------------------------------------------------------+
| Draw scaled image           | See draw image                                            |
+-----------------------------+-----------------------------------------------------------+
```

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 26]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
+
+The following table describes the compatibility versions between the C module and the MicroEJ UI Packs:
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst:345]:
```suggestion
The following table describes the version compatibility between the C module and the MicroEJ UI Packs:
```

File: VEEPortingGuide/uiReleaseNotes.rst [hunk 27]
```diff
@@ -238,6 +238,138 @@ The following table describes the compatibility versions between the C modules a
...
++----------------+-----------------+
+| C Module Range | UI Pack Range   |
++================+=================+
+| [3.0.0-5.0.1]  | [13.3.0-13.4.1] |
++----------------+-----------------+
+| [1.0.0-2.0.0]  | [13.1.0-13.3.0] |
++----------------+-----------------+
+
+
+The following table describes the compatibility versions between the C module and the VG-Lite libraries (officially supported):
```
Comment (microej-sh) [inline VEEPortingGuide/uiReleaseNotes.rst:356]:
```suggestion
The following table describes the version compatibility between the C module and the VG-Lite libraries (officially supported):
```

**Review by microej-sh — CHANGES_REQUESTED** [review]

**Review by microej-sh — APPROVED** [review]
